sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberUtils.scala

```diff
@@ -0,0 +1,189 @@
+    // or "###,###".
+    normalizedFormat.stripPrefix(commaSignStr).stripSuffix(commaSignStr)
+  }
+
+  private def isSign(c: Char): Boolean = {
+    Set(pointSign, commaSign, minusSign, dollarSign).contains(c)
+  }
+
+  private def transform(format: String): String = {
+    if (format.contains(minusSign)) {
+      val positiveFormatString = format.replaceAll("-", "")
+      s"$positiveFormatString;$format"
+    } else {
+      format
+    }
+  }
+
+  private def check(normalizedFormat: String, numberFormat: String) = {
+    def invalidSignPosition(format: String, c: Char): Boolean = {
+      val signIndex = format.indexOf(c)
+      signIndex > 0 && signIndex < format.length - 1
+    }
+
+    if (normalizedFormat.count(_ == pointSign) > 1) {
+      throw QueryCompilationErrors.multipleSignInNumberFormatError(
+        s"'$letterPointSign' or '$pointSign'", numberFormat)
+    } else if (normalizedFormat.count(_ == minusSign) > 1) {
+      throw QueryCompilationErrors.multipleSignInNumberFormatError(
+        s"'$letterMinusSign' or '$minusSign'", numberFormat)
+    } else if (normalizedFormat.count(_ == dollarSign) > 1) {
+      throw QueryCompilationErrors.multipleSignInNumberFormatError(s"'$dollarSign'", numberFormat)
+    } else if (invalidSignPosition(normalizedFormat, minusSign)) {
+      throw QueryCompilationErrors.nonFistOrLastCharInNumberFormatError(
+        s"'$letterMinusSign' or '$minusSign'", numberFormat)
+    } else if (invalidSignPosition(normalizedFormat, dollarSign)) {
+      throw QueryCompilationErrors.nonFistOrLastCharInNumberFormatError(
+        s"'$dollarSign'", numberFormat)
+    }
+  }
+
+  /**
+   * Convert string to numeric based on the given number format.
+   * The format can consist of the following characters:
+   * '9':  digit position (can be dropped if insignificant)
+   * '0':  digit position (will not be dropped, even if insignificant)
+   * '.':  decimal point (only allowed once)
+   * ',':  group (thousands) separator
+   * 'S':  sign anchored to number (uses locale)
+   * 'D':  decimal point (uses locale)
+   * 'G':  group separator (uses locale)
+   * '$':  specifies that the input value has a leading $ (Dollar) sign.
+   *
+   * @param input the string need to converted
+   * @param numberFormat the given number format
+   * @return decimal obtained from string parsing
+   */
+  def parse(input: UTF8String, numberFormat: String): Decimal = {
+    val normalizedFormat = normalize(numberFormat)
+    check(normalizedFormat, numberFormat)
+
+    val precision = normalizedFormat.filterNot(isSign).length
+    val formatSplits = normalizedFormat.split(pointSign)
+    val scale = if (formatSplits.length == 1) {
+      0
+    } else {
+      formatSplits(1).filterNot(isSign).length
+    }
+    val transformedFormat = transform(normalizedFormat)
+    val numberFormatInstance = NumberFormat.getInstance()
+    val numberDecimalFormat = numberFormatInstance.asInstanceOf[DecimalFormat]
```

Comment:
is it always true? can you refer to some JDK doc?

Reply:
Yes. The default choice is `NUMBERSTYLE`, it will return `DecimalFormat` indeed.
```
public final static NumberFormat getInstance() {
    return getInstance(Locale.getDefault(Locale.Category.FORMAT), NUMBERSTYLE);
}
```
Let's add assert.